File: populate_movie.py
import requests
import psycopg2
from dotenv import load_dotenv
import os

load_dotenv()

DB_HOST = os.getenv('DB_HOST')
DB_NAME = os.getenv('DB_NAME')
DB_USER = os.getenv('DB_USER')
DB_PASSWORD = os.getenv('DB_PASSWORD')
RAPIDAPI_KEY = os.getenv("RAPIDAPI_KEY")
RAPIDAPI_HOST = os.getenv("RAPIDAPI_HOST")

# Connect to PostgreSQL database
conn = psycopg2.connect(
    host=DB_HOST,
    database=DB_NAME,
    user=DB_USER,
    password=DB_PASSWORD
)

# This file fetches the details from IMDBot and WatchMode in order to populate the movie table 
# Info obtained from IMDBot: title, plot, viewer_rating, release_year, imdb_id
# Info obtained from WatchMode: tmdb_id, watchmode_id


# Fetch movie details from IMDBot API
def fetch_movie_details():
    url = "https://search.imdbot.workers.dev/?tt=tt1099212"  # for Twilight - part 2 & 3 movie
    response = requests.get(url)
    
    if response.status_code != 200:
        print("Error fetching IMDb data:", response.status_code)
        return None, None, None, None, None

    json_response = response.json()
    short_data = json_response.get("short", {})
    title = short_data.get("name")
    plot = short_data.get("description")
    rating = short_data.get("aggregateRating", {}).get("ratingValue")
    viewers_rating = float(rating) if rating else None
    release_year = int(short_data.get("datePublished", "")[:4]) if short_data.get("datePublished") else None
    imdb_id = json_response.get("imdbId")
    print(f"short data", imdb_id)
    return title, plot, viewers_rating, release_year, imdb_id

# Fetch Watchmode IDs using imdb_id obtained from IMDBot
def fetch_watchmode_movie_details(imdb_id):
    url = f"https://watchmode.p.rapidapi.com/search/?search_field=imdb_id&search_value={imdb_id}&types=movie"
    
    headers = {
        'x-rapidapi-key': RAPIDAPI_KEY,
        'x-rapidapi-host': 'watchmode.p.rapidapi.com'
    }

    response = requests.get(url, headers=headers)

    if response.status_code != 200:
        print(f"Error fetching Watchmode data: {response.status_code}")
        return None, None

    json_response = response.json()
    title_results = json_response.get("title_results")

    if not title_results:
        print("No title results found in Watchmode for this IMDb ID.")
        return None, None

    tmdb_id = title_results[0].get("tmdb_id", None)
    watchmode_id = title_results[0].get("id", None)

    print("TMDB ID:", tmdb_id)
    print("Watchmode ID:", watchmode_id)
    
    return tmdb_id, watchmode_id

# Insert movie details into the database
def insert_movie_details(title, plot, viewers_rating, release_year, imdb_id, tmdb_id, watchmode_id):
    with conn.cursor() as cur:
        try:
            cur.execute("""
                INSERT INTO movie (title, plot, viewers_rating, release_year, imdb_id, tmdb_id, watchmode_id)
                VALUES (%s, %s, %s, %s, %s, %s, %s)
            """, (title, plot, viewers_rating, release_year, imdb_id, tmdb_id, watchmode_id))
            conn.commit()
            print("Movie details inserted successfully.")
        except Exception as e:
            print(f"Error inserting movie details: {e}")
            conn.rollback()

# Fetch and insert details from both APIs
def populate_all_details_from_apis():
    title, plot, viewers_rating, release_year, imdb_id = fetch_movie_details()
    
    if not imdb_id:
        print("IMDb ID not found, skipping Watchmode fetch.")
        return

    tmdb_id, watchmode_id = fetch_watchmode_movie_details(imdb_id)
    
    if title:
        insert_movie_details(title, plot, viewers_rating, release_year, imdb_id, tmdb_id, watchmode_id)

populate_all_details_from_apis()

File: populate_movie.js
const axios = require('axios');
const { Client } = require('pg');

// Set up your database connection
const client = new Client({
  host: 'localhost',
  port: 5432,
  user: 'postgres',
  password: 'jiro', // Replace with your password
  database: 'A2_363', // Replace with your DB name
});

// Your RapidAPI key for the Watchmode API
const RAPIDAPI_KEY = 'd044bde95amshea8beb39be77448p19cb8ejsn44384da3c58b'; // Replace with your actual API key

// Connect to the database
client.connect()
  .then(() => console.log('Connected to PostgreSQL'))
  .catch(err => console.error('Error connecting to PostgreSQL', err.stack));

// API URL to fetch movie details
const imdbApiUrl = 'https://search.imdbot.workers.dev/?tt=tt0107688';

axios.get(imdbApiUrl)
  .then(async (response) => {
    const jsonResponse = response.data;
    const shortData = jsonResponse.short;

    // Fetching movie details
    const movieTitle = shortData.name || '';
    const plot = shortData.description || '';
    const rating = shortData.aggregateRating?.ratingValue || null;
    const viewersRating = rating ? parseFloat(rating) : null;
    const releaseYear = shortData.datePublished ? parseInt(shortData.datePublished.slice(0, 4)) : null;
    const imdbId = jsonResponse.imdbId || null;

    // Fetch tmdb_id and watchmode_id using the Watchmode API
    let tmdbId = null;
    let watchmodeId = null;

    if (imdbId) {
      // Use the URL format you specified
      const watchmodeApiUrl = `https://watchmode.p.rapidapi.com/search/?search_field=imdb_id&search_value=${imdbId}&types=movie`;
      const headers = {
        'x-rapidapi-key': RAPIDAPI_KEY,
        'x-rapidapi-host': 'watchmode.p.rapidapi.com',
      };

      try {
        const watchmodeResponse = await axios.get(watchmodeApiUrl, { headers });
        const titleResults = watchmodeResponse.data.title_results;

        if (titleResults && titleResults.length > 0) {
          // Extracting tmdb_id and watchmode_id from the first result
          tmdbId = titleResults[0]?.tmdb_id || null;
          watchmodeId = titleResults[0]?.id || null;
        }
      } catch (err) {
        console.error('Error fetching data from Watchmode API:', err.message);
      }
    }

    // Insert movie data into the movie table (simplified without content rating, country, or language)
    try {
      const checkMovieQuery = `SELECT id FROM movie WHERE tmdb_id = $1 LIMIT 1`;
      const checkResult = await client.query(checkMovieQuery, [tmdbId]);

      if (checkResult.rows.length === 0) {
        const insertMovieQuery = `
          INSERT INTO movie 
            (title, plot, viewers_rating, release_year, 
             watchmode_id, tmdb_id, imdb_id) 
           VALUES 
            ($1, $2, $3, $4, $5, $6, $7)
           ON CONFLICT (tmdb_id) DO NOTHING;
        `;

        const values = [
          movieTitle,
          plot,
          viewersRating,
          releaseYear,
          watchmodeId,
          tmdbId,
          imdbId,
        ];

        const result = await client.query(insertMovieQuery, values);

        if (result.rowCount > 0) {
          console.log(`Inserted movie: ${movieTitle}`);
        }
      } else {
        console.log(`Movie already exists: ${movieTitle}`);
      }

    } catch (err) {
      console.error('Error inserting movie:', err.stack);
    }
  })
  .catch((error) => {
    console.error('Error fetching data from IMDb API:', error.message);
  })
  .finally(() => {
    client.end();
  });
